keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,141 @@
+
+#### iptables NOTRACK
+
+NOTRACK rules are added for connections to and from the nodelocal dns service ip. Additional rules in FILTER table to whitelist these connections, since the INPUT and OUTPUT chains have a default DROP policy.
+
+A dnscache nanny, similar to the [dnsmasq nanny](https://github.com/kubernetes/dns/tree/master/pkg/dnsmasq) will create the dummy interface and iptables rules. The nanny gets the nodelocal dns service ip address by querying kube-dns for the service name. The nanny will also periodically ensure that the iptables rules are present. The resource usage for periodic iptables check needs to be measured. We can reduce the frequency of the checks by having the nanny query the dns cache and checking/adding rules only if the query fails.
+
+
+[Proposal presentation](https://docs.google.com/presentation/d/1c43cZqbVhGAlw3dSNQIOGuvQmDfKaA2yiAPRoYpa6iY), also shared at the sig-networking meeting on 2018-10-04
+
+Slide 5 has a diagram showing how the new dns cache fits in.
+
+#### Choice of caching agent
+
+The current plan is to run [Unbound dns server](https://www.nlnetlabs.nl/projects/unbound/about/) by default, based on these benchmark[ tests](https://github.com/kubernetes/perf-tests/tree/master/dns).
+
+Will rerun benchmark tests with coreDNS after SO_REUSEPORT support and measure QPS.
+
+Tests were run on a 1.9.7 cluster with 2 nodes, using Unbound 1.7.3 and coreDNS 1.2. Unbound QPS was 3600(caching)/3400(no caching), coreDNS 950(caching)/500(no caching)
+
+
+Based on the prototype/test results, these are the recommended defaults: 
+CPU request: 50m
+Memory Limit : 25m  
+Resource usage and QPS to be recomputed after metrics collection has been added.
+
+#### Metrics
+
+Unbound metrics can be extracted from the program unbound-control. This can be run from the nanny container.
+Unbound-control-setup is required to setup keys/certs for SSL.
+
+
+### Risks and Mitigations
+
+* Having the pods query the nodelocal cache introduces a single point of failure.   
+This is mitigated by assigning a PriorityClass of "system-node-critical", so it is always running. The nanny process will periodically check if the dns agent is running.
+Another option is to populate both the nodelocal cache ip address and kube-dns ip address in resolv.conf. The default behavior will be to try the nameservers in the order they are listed. From the man page of [resolv.conf](https://linux.die.net/man/5/resolv.conf) - "If there are multiple servers, the resolver library queries them in the order listed".
+
+
```

Comment:
Can we add a subsection on how this is to be rolled out?

Also, we would need to adjust the kube-dns scaling parameters I would imagine when this is deployed?

Reply:
Added.

Yes, we would need to adjust autoscaling parameters and also maybe resources to kube-dns and dnsmasq. I was hoping to get some more data once the feature is in alpha and then determine changes to the parameters.